hbase-client/src/main/java/org/apache/hadoop/hbase/client/RawAsyncTableImpl.java

```diff
@@ -796,75 +804,165 @@ private boolean locateFinished(RegionInfo region, byte[] endKey, boolean endKeyI
 
   private <S, R> void coprocessorServiceUntilComplete(Function<RpcChannel, S> stubMaker,
     ServiceCaller<S, R> callable, PartialResultCoprocessorCallback<S, R> callback,
-    AtomicBoolean locateFinished, AtomicInteger unfinishedRequest, RegionInfo region, Span span) {
+    MultiRegionCoprocessorServiceProgress<R> progress, RegionInfo region,
+    AtomicBoolean claimedRegionAssignment, Span span) {
     addListener(coprocessorService(stubMaker, callable, region, region.getStartKey()), (r, e) -> {
       try (Scope ignored = span.makeCurrent()) {
-        if (e != null) {
-          callback.onRegionError(region, e);
+        if (e instanceof RegionNoLongerExistsException) {
+          RegionInfo newRegion = ((RegionNoLongerExistsException) e).getNewRegionInfo();
+          if (progress.markRegionReplacedExistingAndCheckNeedsToBeRestarted(newRegion)) {
```

Comment:
When a region is merged, this is an optimization to avoid sending multiple requests to the new region